daemon/devices_nvidia_linux.go

```diff
@@ -117,11 +114,86 @@ func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
 	return nil
 }
 
+// getRequestedDeviceIDs returns the list of requested devices by ID based on
+// the device request.
+func getRequestedDevicesIDs(req container.DeviceRequest) []string {
+	switch {
+	case len(req.DeviceIDs) > 0:
+		return req.DeviceIDs
+	case req.Count > 0:
+		return countToDevices(req.Count)
+	case req.Count < 0:
+		return []string{"all"}
+	case req.Count == 0:
+		return nil
+	}
+	return nil
+}
+
 // countToDevices returns the list 0, 1, ... count-1 of deviceIDs.
-func countToDevices(count int) string {
+func countToDevices(count int) []string {
 	devices := make([]string, count)
 	for i := range devices {
 		devices[i] = strconv.Itoa(i)
 	}
-	return strings.Join(devices, ",")
+	return devices
+}
+
+// A cdiDeviceInjector is used to map regular device requests to CDI device
+// requests.
+type cdiDeviceInjector struct {
+	defaultCDIDeviceKind                 string
+	fallbackToNVIDIAContainerRuntimeHook bool
+}
+
+// injectDevices converts an incoming device request to a request for devices
+// using CDI.
+// The requested device IDs are converted to CDI device names if required using
+// the specified default kind.
+func (i *cdiDeviceInjector) injectDevices(s *specs.Spec, dev *deviceInstance) error {
+	var cdiDeviceIDs []string
+	for _, deviceID := range getRequestedDevicesIDs(dev.req) {
+		cdiDeviceIDs = append(cdiDeviceIDs, i.normalizeDeviceID(deviceID))
+	}
+	if len(cdiDeviceIDs) == 0 {
+		return nil
+	}
+
+	// If the cdi device driver is not available then we return an error.
+	cdiDeviceDriver := deviceDrivers["cdi"]
+	if cdiDeviceDriver == nil {
+		if i.fallbackToNVIDIAContainerRuntimeHook {
+			return setNvidiaGPUs(s, dev)
+		}
+		return incompatibleDeviceRequest{dev.req.Driver, dev.req.Capabilities}
+	}
+
```

Comment:
One question would be whether we wanted to check whether the requested devices are known at this stage and only then invoke the `cdiDeviceDriver`. Alternatively, we could simplyfy the fallback logic to always fallback to the `setNvidiaGPUs` handler for all errors.